### ISSUE TITLE
Migration guide for new version package

### ISSUE THREAD

**Issue by bambinoua** [baumths/flutter_tree_view#40]
It is good to have a new version of this package? And what about migration guide because of tons of breaking changes? I've updated version and I am horrified how many code I need to change! 
For example where `TreeViewTheme` or what do I need to use now?

**Comment by bambinoua** [comment]
I cannot migrate my project to new version! You changed base classes, you change everything! How to specify `nodeHeight`? Where `expandAll` and `collapseAll` method? How to migrate??? Why such new version?

**Comment by baumths** [comment]
TL;DR

Please treat v1.0 as a new package, avoid attempting to migrate.

As for the `TreeViewTheme`, checkout the new [TreeIndentation][https://pub.dev/documentation/flutter_fancy_tree_view/latest/flutter_fancy_tree_view/TreeIndentation-class.html] widget and the [IndentGuide][https://pub.dev/documentation/flutter_fancy_tree_view/latest/flutter_fancy_tree_view/IndentGuide-class.html] configuration.

---

Because of the many new features I have planned for the future of this package, I had to rewrite the entire API from scratch. It did indeed create a lot of breaking changes and I'm truly sorry about that. But those changes were all necessary for the future minor releases of 1.0.

The new API is more robust and versatile and will payoff in the future.

I currently do not have the time to write a migration guide, that's why I'm asking to treat this new version as a whole new package instead of migrating. If you feel that upgrading to 1.0 is a big undertaking I'd advise you to not migrate and keep using the 0.x version until either you have some spare time to learn and apply the new APIs or for me to get some spare time and work on a migration guide (which is probably not happening anytime soon as I have very limited time to code due to my occupation).

**Comment by baumths** [comment]
> I cannot migrate my project to new version! You changed base classes, you changed everything! How to specify `nodeHeight`? Where `expandAll` and `collapseAll` methods? How to migrate??? Why such new version?

`nodeHeight` is not needed anymore, you can size your node widgets as you will.

`expandAll` and `collapseAll` were replaced by `expandCascading` and `collapseCascading`.

**Comment by baumths** [comment]
Please be nice to open source maintainers 😅

**Comment by bambinoua** [comment]
> Please treat v1.0 as a new package, avoid attempting to migrate.

Just imaging, you upgrades to new flutter and then... oops all `Widget`s and any controllers have different prototypes etc.  One functionality disappeared, other functionality renamed and you do not know how and what to use (because no migration docs or at lease `@deprecated` annotations). You see tons of errors and you do not know how to fix them. So the rational question why do you need such flutter upgrade?

I use your pacakge in many places and now my project is broken and stuck because I am not sure if I can trust this new package and the global business logic was not drastically changed.

Yes, the simple way is just to rollback to old version but why do I (or other person) needs this package then?

> I currently do not have the time to write a migration guide

Yeah... :)

**Comment by bambinoua** [comment]
> Please be nice to open source maintainers 😅

Please be careful to users which you (open source maintainers) suggest their products for. If open source maintainers don't care on their customer so maybe no need to suggest open product?

**Comment by bambinoua** [comment]
@amorenew I understand. I upgraded packages (because of new version) almost each day and I **never** had big problems (just minor mentioned deprecations or even no changes). I tied with this package in a few places of my project and now I have a huge problem because this version is completely different! 

Yes I can to rollback but how can I be sure that old version will not have bugs? And if it will have bugs how they will be fixed if there is already new codebase?

I understand that this is an open project but developer must have a definite responsibility if he made his project public.

**Comment by bambinoua** [comment]
@baumths what about my questions about `expandedNodes`, `lineThickness` and `lineColor`?

**Comment by bambinoua** [comment]
@baumths how can I get information about expanded nodes? I used `expandedNodes` list for my needs for restoration expanded nodes. Now I see this property is private. Is it possible to make it public? Or how can I get such data now?

P.S. Sorry for emotions.

**Comment by baumths** [comment]
> @baumths how can I get information about expanded nodes? I used `expandedNodes` list for my needs for restoration expanded nodes. Now I see this property is private. Is it possible to make it public? Or how can I get such data now?

The new way of keeping track of expanded nodes is to extend the `TreeController` class and override the `getExpansionState` and `setExpansionState` to intercept the value changes (calling `super.` to keep the default behavior), both methods might be all you need to keep track of the expanded nodes.

Try it out and let me know if it was enough otherwise I might make `expandedNodes` public again (I made it private since it is just a default implementation for the `TreeController` to avoid making the controller `abstract` and have users implement the interface) I expected users to override this behavior depending on the use case.

About `visibleNodes` that you asked earlier, you can achieve the same behavior with the following snippet of code:
```dart
final TreeController<T> controller = ...;
final List<T> visibleNodes = []
controller.depthFirstTraversal(
  onTraverse: (TreeEntry<T> entry) {
    visibleNodes.add(entry.node);
  },
);
```
> The `depthFirstTraversal` method also takes in an optional `descendCondition` predicate that can be used to limit the reach of the traversal, by default it checks if the provided tree node is expanded which will result in traversing the tree in the same order as the previous `visibleNodes` property.

Sorry for the delayed response, busy day 😅.

**Comment by gslender** [comment]
Yep, totally understand the issue outlined here about making major wholesale changes to the API, but can you maybe include some approx direction on how to migrate a simple TreeView example that used things like TreeNode and a few of the common trimmings into the new API.

I'd also encourage you to put that guideance (about needing to start again) in the CHANGELOG, as I always read them but didn't get that degree of breaking change out of the notes. It wasn't until I found way too many things missing that I realsied I needed to read this issue post, and then back out of my changes - all and all, not very nice, but understandable. I think you may have just failed to really explain the extent of the changes made, so I'd only ask that maybe a clearer upfront notice of that is provided to avoid others like me from wasting time trying to migrate when realising the refactoring is signifcant, if not impactical in some parts without changing the architecture of your existing app.

**Comment by bambinoua** [comment]
> Try it out and let me know if it was enough otherwise I might make expandedNodes public again
Thanks. I will. Maybe this a good idea because in my case (and in old version) I used nodes ids (for persist) and I don't need to store whole nodes.

**Comment by bambinoua** [comment]
@baumths Finally I can compile my project commenting all my functionality (persist expand/collapse status, expanding/collapsing all tree) but now I see only my roots and I don't see children. How can I force to expand the tree? My tree data is populated.

**Comment by bambinoua** [comment]
@baumths I tried to make my tree totally expanded so I did
```dart
final controller = FancyTreeController(nodes)..expandAll();
```
```dart
/// My extended class
class FancyTreeController extends TreeController {
  void expandAll() {
    expandCascading(roots); // Here I wanted to expand all roots
  }
}
```
What did I wrong?
